From ea04d983c6ba94da70cb2eb646baca2e69a7fee7 Mon Sep 17 00:00:00 2001
From: Gauthier Picalausa <mc2m@uplinklabs.net>
Date: Sat, 12 Jul 2014 16:08:33 +0200
Subject: [PATCH] Ensure the usage of the correct run-time library

---
 src/msvcbuild.bat | 12 ++++++++----
 1 file changed, 8 insertions(+), 4 deletions(-)

diff --git a/src/msvcbuild.bat b/src/msvcbuild.bat
index 9160e0f..7301e68 100644
--- a/src/msvcbuild.bat
+++ b/src/msvcbuild.bat
@@ -14,6 +14,8 @@
 @if not defined INCLUDE goto :FAIL
 
 @setlocal
+@set LJRTDLIB=/MD
+@set LJRTLIB=/MT
 @set LJCOMPILE=cl /nologo /c /O2 /W3 /D_CRT_SECURE_NO_DEPRECATE
 @set LJLINK=link /nologo
 @set LJMT=mt /nologo
@@ -67,22 +69,24 @@ buildvm -m folddef -o lj_folddef.h lj_opt_fold.c
 @shift
 @set LJCOMPILE=%LJCOMPILE% /Zi
 @set LJLINK=%LJLINK% /debug
+@set LJRTLIB=%LJRTLIB%d
+@set LJRTDLIB=%LJRTDLIB%d
 :NODEBUG
 @if "%1"=="amalg" goto :AMALGDLL
 @if "%1"=="static" goto :STATIC
-%LJCOMPILE% /MD /DLUA_BUILD_AS_DLL lj_*.c lib_*.c
+%LJCOMPILE% %LJRTDLIB% /DLUA_BUILD_AS_DLL lj_*.c lib_*.c
 @if errorlevel 1 goto :BAD
 %LJLINK% /DLL /out:%LJDLLNAME% lj_*.obj lib_*.obj
 @if errorlevel 1 goto :BAD
 @goto :MTDLL
 :STATIC
-%LJCOMPILE% lj_*.c lib_*.c
+%LJCOMPILE% %LJRTLIB% lj_*.c lib_*.c
 @if errorlevel 1 goto :BAD
 %LJLIB% /OUT:%LJLIBNAME% lj_*.obj lib_*.obj
 @if errorlevel 1 goto :BAD
 @goto :MTDLL
 :AMALGDLL
-%LJCOMPILE% /MD /DLUA_BUILD_AS_DLL ljamalg.c
+%LJCOMPILE% %LJRTDLIB% /DLUA_BUILD_AS_DLL ljamalg.c
 @if errorlevel 1 goto :BAD
 %LJLINK% /DLL /out:%LJDLLNAME% ljamalg.obj lj_vm.obj
 @if errorlevel 1 goto :BAD
@@ -90,7 +94,7 @@ buildvm -m folddef -o lj_folddef.h lj_opt_fold.c
 if exist %LJDLLNAME%.manifest^
   %LJMT% -manifest %LJDLLNAME%.manifest -outputresource:%LJDLLNAME%;2
 
-%LJCOMPILE% luajit.c
+%LJCOMPILE% %LJRTLIB% luajit.c
 @if errorlevel 1 goto :BAD
 %LJLINK% /out:luajit.exe luajit.obj %LJLIBNAME%
 @if errorlevel 1 goto :BAD
-- 
1.8.3.msysgit.0

